### PR TITLE
[PAGOPA-2109] custom client

### PR DESCRIPTION
**Pull request by jacopocarlini**
#### List of Changes

♻️ custom apiclient to hanlde appException and Retry

#### Motivation and Context

↬ retry HTTP call to external services

#### How Has This Been Tested?

✋ manually
🤖 junit

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/java/it/gov/pagopa/wispconverter/config/client/custom/ApiConfigCacheApiClient.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    private int handleRetry(AppException ex, Throwable cause, int attempts) {
+        if (cause instanceof HttpServerErrorException
+                || ((HttpClientErrorException) cause)
+                .getStatusCode()
+                .equals(HttpStatus.TOO_MANY_REQUESTS)) {
+            attempts++;
+            if (attempts < super.getMaxAttemptsForRetry()) {
+                try {
+                    Thread.sleep(super.getWaitTimeMillis());
```
Comment (andrea-deri) [inline src/main/java/it/gov/pagopa/wispconverter/config/client/custom/ApiConfigCacheApiClient.java:57]: How long is the waiting time used? Can be defined as parameterized variable (like `max-retry`) in order to control it?

**Review by andrea-deri — CHANGES_REQUESTED** [review]
In general, it is a good proposal! 
I have doubts about a few points to which I have added comments. A fundamental question: has this new handling been explicitly tested on a service (e.g. IUV-Generator) that times out?

File: src/main/java/it/gov/pagopa/wispconverter/config/client/custom/ApiConfigCacheApiClient.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                                           MultiValueMap<String, Object> formParams,
+                                           List<MediaType> accept,
+                                           MediaType contentType,
+                                           String[] authNames,
+                                           ParameterizedTypeReference<T> returnType)
+            throws RestClientException {
+
+        int attempts = 0;
+        ResponseEntity<T> response = null;
+        while (attempts < super.getMaxAttemptsForRetry()) {
```
[the remark this answers is not in the document]
Comment (jacopocarlini) [reply src/main/java/it/gov/pagopa/wispconverter/config/client/custom/ApiConfigCacheApiClient.java:38]: I missed the `break` operation 😅

File: src/main/java/it/gov/pagopa/wispconverter/config/client/custom/ApiConfigCacheApiClient.java (hunk 1 above)
Comment (jacopocarlini) [reply src/main/java/it/gov/pagopa/wispconverter/config/client/custom/ApiConfigCacheApiClient.java:57]: ok, I can add the parameters. The default value is 10 millisec

**Comment by jacopocarlini** [comment]
> In general, it is a good proposal! I have doubts about a few points to which I have added comments. A fundamental question: has this new handling been explicitly tested on a service (e.g. IUV-Generator) that times out?

Yes, I did.

File: src/main/java/it/gov/pagopa/wispconverter/config/client/custom/CheckoutApiClient.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                attempts = handleRetry(ex, attempts);
+            }
+        }
+        return response;
+    }
+
+    private int handleAppException(AppException ex, int attempts) {
+        Throwable cause = ex.getCause();
+        if (cause instanceof RuntimeException runtimeException) {
+            if (cause instanceof HttpServerErrorException || ((HttpClientErrorException) cause).getStatusCode().equals(HttpStatus.TOO_MANY_REQUESTS)) {
```
Comment (FedericoRuzzier) [inline src/main/java/it/gov/pagopa/wispconverter/config/client/custom/CheckoutApiClient.java:51]: Add log to explicit the error which raised the retry

File: src/main/java/it/gov/pagopa/wispconverter/config/client/custom/ApiConfigCacheApiClient.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            }
+        }
+        return response;
+    }
+
+    private int handleAppException(AppException ex, int attempts) {
+        Throwable cause = ex.getCause();
+        if (cause instanceof RuntimeException runtimeException) {
+            if (cause instanceof HttpServerErrorException || ((HttpClientErrorException) cause).getStatusCode().equals(HttpStatus.TOO_MANY_REQUESTS)) {
+                attempts = handleRetry(runtimeException, attempts);
```
Comment (FedericoRuzzier) [inline src/main/java/it/gov/pagopa/wispconverter/config/client/custom/ApiConfigCacheApiClient.java:52]: Ditto log

**Pull request merged**